virtcontainers/netmon/netmon.go

```diff
@@ -0,0 +1,531 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/kata-containers/agent/protocols/grpc"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	netmonName    = "kata-netmon"
+	netmonVersion = "0.0.1"
+
+	kataCLIAddIfaceCmd   = "add-interface"
+	kataCLIDelIfaceCmd   = "del-interface"
+	kataCLIUpdtRoutesCmd = "update-routes"
+
+	kataSuffix = "kata"
+
+	netlinkFamily = netlink.FAMILY_V4
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
+	storageDirPerm    = 0750
+	sharedFile        = "shared.json"
+	storageFilePerm   = 0640
+)
+
+type netmonParams struct {
+	sandboxID   string
+	runtimePath string
+	netnsPath   string
+	debug       bool
+}
+
+type netmon struct {
+	netmonParams
+
+	storagePath string
+	sharedFile  string
+
+	netIfaces map[int]grpc.Interface
+
+	linkUpdateCh chan netlink.LinkUpdate
+	linkDoneCh   chan struct{}
+
+	rtUpdateCh chan netlink.RouteUpdate
+	rtDoneCh   chan struct{}
+
+	netHandler *netlink.Handle
+}
+
+func printVersion() {
+	fmt.Printf("%s version %s\n", netmonName, netmonVersion)
+}
+
+func parseOptions() netmonParams {
+	var version bool
+
+	params := netmonParams{}
+
+	flag.BoolVar(&params.debug, "d", false, "enable debug mode")
+	flag.BoolVar(&version, "v", false, "display program version and exit")
+	flag.StringVar(&params.sandboxID, "s", "", "sandbox id (required)")
+	flag.StringVar(&params.runtimePath, "r", "", "runtime path (required)")
```

Comment:
IMO, one benefit of calling the kata API directly is that if you modularize it properly, we can create a goroutine to monitor the netns in containerd kata shim and that presents no dependency on the kata-runtime cli, which is good from containerd kata shim perspective.

How about a structure like this:
- virtcontainers/netnsmon: a package that calls kata API to do netns monitoring
- cli/netmon: a command line built on top of virtcontainers/netnsmon. Other kata-runtime sub-command can invoke it to create a daemon to monitor the netns
- for containerd kata shim, it can create a goroutine that calls virtcontainers/netnsmon to monitor the netns when it wants to.

In both kata-runtime and containerd kata shim case, virtcontainers do not invoke netnsmon but let the callers decide and drive the procedure.

WDTY?

cc @gnawux @lifupan